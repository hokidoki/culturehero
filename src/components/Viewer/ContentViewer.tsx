import React from 'react';
import ContentViewerStyles from './ContentViewer.styles';
import ContentVeiwerProps from './ContentViewer.types';

export default function ContentViewer({
  id,
  createdAt,
  article,
  deleteButtonOnClick,
  editButtonOnClick,
}: ContentVeiwerProps) {
  return (
    <ContentViewerStyles.Layout>
      <ContentViewerStyles.ViewerHeader>
        <ContentViewerStyles.CreatedAt>
          {createdAt}
        </ContentViewerStyles.CreatedAt>
        <ContentViewerStyles.Controlls>
          <ContentViewerStyles.ControlButton
            color="plain"
            onClick={() => editButtonOnClick(id)}
          >
            수정
          </ContentViewerStyles.ControlButton>
          <ContentViewerStyles.ControlButton
            color="red"
            onClick={() => deleteButtonOnClick(id)}
          >
            삭제
          </ContentViewerStyles.ControlButton>
        </ContentViewerStyles.Controlls>
      </ContentViewerStyles.ViewerHeader>
      <ContentViewerStyles.Article
        dangerouslySetInnerHTML={{
          __html: article.replace(/(?:\r\n|\r|\n)/g, '<br>'),
        }}
      />
    </ContentViewerStyles.Layout>
  );
}
