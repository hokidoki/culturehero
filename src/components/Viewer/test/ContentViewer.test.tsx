import { fireEvent, render, waitFor } from '@render';
import ContentViewer from '../ContentViewer';
import ContentVeiwerProps from '../ContentViewer.types';
import { isNull } from '@fxts/core';

describe('ContentViewer', () => {
  it('match snapshot ', () => {
    const props: ContentVeiwerProps = {
      article: 'ARTICLE',
      createdAt: '2020-02-01',
      id: 'A',
      title: 'TITLE',
      editButtonOnClick: () => {},
      deleteContent: () => Promise.resolve(''),
    };
    const utils = render(<ContentViewer {...props} />);
    expect(utils.container).toMatchSnapshot();
  });

  it('수정 버튼을 클릭하면 editButtonOnClick 이 호출 된다.', () => {
    const props: ContentVeiwerProps = {
      article: 'ARTICLE',
      createdAt: '2020-02-01',
      id: 'A',
      title: 'TITLE',
      editButtonOnClick: jest.fn(),
      deleteContent: () => Promise.resolve(''),
    };
    const utils = render(<ContentViewer {...props} />);
    const putBtn = utils.getByText('수정');
    if (isNull(putBtn)) throw new Error('Missing element');

    fireEvent.click(putBtn);

    expect(props.editButtonOnClick).toBeCalled();
  });

  it('삭제 버튼을 클릭하면 deleteContent가 호출되며 각 버튼들이 disabled상태가 된다.', async () => {
    Object.defineProperty(window, 'confirm', { value: () => true });
    const props: ContentVeiwerProps = {
      article: 'ARTICLE',
      createdAt: '2020-02-01',
      id: 'A',
      title: 'TITLE',
      editButtonOnClick: jest.fn(),
      deleteContent: jest.fn().mockResolvedValue(''),
    };
    const utils = render(<ContentViewer {...props} />);
    const putBtn = utils.getByText('수정');
    const deleteBtn = utils.getByText('삭제');
    if (isNull(deleteBtn)) throw new Error('Missing element');

    fireEvent.click(deleteBtn);

    expect(props.deleteContent).toBeCalled();
    await waitFor(() => {
      expect(putBtn.getAttribute('disabled')).toBe('');
      expect(deleteBtn.getAttribute('disabled')).toBe('');
    });
  });
});
