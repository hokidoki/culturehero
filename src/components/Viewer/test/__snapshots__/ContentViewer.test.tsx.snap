// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentViewer match snapshot  1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  min-height: 700px;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
}

.c1 {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  height: 50px;
  width: 100%;
  padding: 0px 10px;
  border-bottom: 1px solid rgb(46,48,51);
}

.c2 {
  color: #bdc1c6;
  font-size: 1rem;
}

.c3 {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  gap: 10px;
}

.c4 {
  background-color: transparent;
  border: none;
  color: #9aa0a6;
}

.c5 {
  background-color: transparent;
  border: none;
  color: palevioletred;
}

.c6 {
  padding: 10px;
  color: #9aa0a6;
  width: 100%;
  flex-grow: 1;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        2020-02-01
      </span>
      <div
        class="c3"
      >
        <button
          class="c4"
          color="plain"
        >
          수정
        </button>
        <button
          class="c5"
          color="red"
        >
          삭제
        </button>
      </div>
    </div>
    <p
      class="c6"
    >
      ARTICLE
    </p>
  </div>
</div>
`;
